src/vmm/src/device_manager/mmio.rs

```diff
@@ -102,39 +103,30 @@ pub struct MMIODeviceInfo {
 /// Manages the complexities of registering a MMIO device.
 pub struct MMIODeviceManager {
     pub(crate) bus: devices::Bus,
-    pub(crate) irq_allocator: IdAllocator,
-    pub(crate) address_allocator: AddressAllocator,
     pub(crate) id_to_dev_info: HashMap<(DeviceType, String), MMIODeviceInfo>,
 }
 
 impl MMIODeviceManager {
     /// Create a new DeviceManager handling mmio devices (virtio net, block).
-    pub fn new(
-        mmio_base: u64,
-        mmio_size: u64,
-        (irq_start, irq_end): (u32, u32),
-    ) -> Result<MMIODeviceManager> {
+    pub fn new() -> Result<MMIODeviceManager> {
         Ok(MMIODeviceManager {
-            irq_allocator: IdAllocator::new(irq_start, irq_end).map_err(Error::AllocatorError)?,
-            address_allocator: AddressAllocator::new(mmio_base, mmio_size)
-                .map_err(Error::AllocatorError)?,
             bus: devices::Bus::new(),
             id_to_dev_info: HashMap::new(),
         })
     }
 
     /// Allocates resources for a new device to be added.
-    fn allocate_new_slot(&mut self, irq_count: u32) -> Result<MMIODeviceInfo> {
-        let irqs = (0..irq_count)
-            .map(|_| self.irq_allocator.allocate_id())
-            .collect::<vm_allocator::Result<_>>()
+    pub(crate) fn allocate_new_slot(
```

Comment:
IMO, this function doesn't need to be public.
```suggestion
    fn allocate_new_slot(
```
Consumers of `MMIODeviceManager` can use device type-specific public methods (like `regsiter_mmio_virtio()`, `register_mmio_serial()`, etc.).
In addition, issue #2014 claims that the current implementation does IRQ allocation and registration in different place in the code and they should be done in one place.
Keeping this method private guarantees both IRQ allocation and registration are done via device type-specific public methods.